From c7619ff7c9c6cf382bc3555c617444b1e408cada Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 21 Jul 2016 19:36:14 +0200
Subject: [PATCH] sched: dovetail: add stage debug check to common scheduling
 paths

Only code running over the in-band stage may reschedule, so checking
for the current stage in might_resched() can detect spurious calling
contexts early.

Likewise, all wake_up*() calls must be issued from the in-band stage,
so check the current stage from the common task wakeup helper too.

This code is compiled in with CONFIG_DEBUG_IRQ_PIPELINE.
---
 include/linux/kernel.h | 8 ++++++--
 kernel/sched/core.c    | 2 ++
 kernel/sched/wait.c    | 2 ++
 3 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/include/linux/kernel.h b/include/linux/kernel.h
index 0c9bc231107f..db575c38e06c 100644
--- a/include/linux/kernel.h
+++ b/include/linux/kernel.h
@@ -14,6 +14,7 @@
 #include <linux/typecheck.h>
 #include <linux/printk.h>
 #include <linux/build_bug.h>
+#include <asm-generic/irq_pipeline.h>
 #include <asm/byteorder.h>
 #include <asm/div64.h>
 #include <uapi/linux/kernel.h>
@@ -201,9 +202,12 @@ struct user;
 
 #ifdef CONFIG_PREEMPT_VOLUNTARY
 extern int _cond_resched(void);
-# define might_resched() _cond_resched()
+# define might_resched() do { \
+		check_inband_stage(); \
+		_cond_resched(); \
+	} while (0)
 #else
-# define might_resched() do { } while (0)
+# define might_resched() check_inband_stage()
 #endif
 
 #ifdef CONFIG_DEBUG_ATOMIC_SLEEP
diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index bae22f25987d..62326bbe00b9 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -3272,6 +3272,8 @@ static inline void schedule_debug(struct task_struct *prev)
 		panic("corrupted stack end detected inside scheduler\n");
 #endif
 
+	check_inband_stage();
+
 	if (unlikely(in_atomic_preempt_off())) {
 		__schedule_bug(prev);
 		preempt_count_set(PREEMPT_DISABLED);
diff --git a/kernel/sched/wait.c b/kernel/sched/wait.c
index fa0f9adfb752..7f5cc7eb17b8 100644
--- a/kernel/sched/wait.c
+++ b/kernel/sched/wait.c
@@ -70,6 +70,8 @@ static int __wake_up_common(struct wait_queue_head *wq_head, unsigned int mode,
 	wait_queue_entry_t *curr, *next;
 	int cnt = 0;
 
+	check_inband_stage();
+
 	lockdep_assert_held(&wq_head->lock);
 
 	if (bookmark && (bookmark->flags & WQ_FLAG_BOOKMARK)) {
-- 
2.16.4

